libs/wal_decoder/src/models.rs

```diff
@@ -25,14 +25,56 @@
 //!                     |--> write to KV store within the pageserver
 
 use bytes::Bytes;
+use pageserver_api::key::CompactKey;
 use pageserver_api::reltag::{RelTag, SlruKind};
+use pageserver_api::value::Value;
 use postgres_ffi::walrecord::{
     XlMultiXactCreate, XlMultiXactTruncate, XlRelmapUpdate, XlReploriginDrop, XlReploriginSet,
     XlSmgrTruncate, XlXactParsedRecord,
 };
 use postgres_ffi::{Oid, TransactionId};
 use utils::lsn::Lsn;
 
+pub enum FlushUncommittedRecords {
+    Yes,
+    No,
+}
+
+/// An interpreted Postgres WAL record, ready to be handled by the pageserver
+pub struct InterpretedWalRecord {
```

Comment:
How do we ensure integrity of the data here? The WAL records have checksums, which are verified during decoding, but we're going to serialize `InterpretedWalRecord` and ship it across the network again -- do we have any further checksumming along the way (e.g. in the network protocol or block formats), or should we consider adding something? With gRPC we would get this for free via TLS. Perhaps we can add a simple header in the network protocol.

Fine to leave this for when we lift this to Safekeeper, but let's make a note of it somewhere so we don't forget.

Reply:
The underlying protocol is TCP, so we do get some checksumming from that, but you make a good point that right now we have application level checksumming. I think it makes sense to preserve that. Let's worry about it when we do the lift to the safekeeper.

Reply:
Updated https://github.com/neondatabase/neon/issues/9336